exasol/driver/websocket/_cursor.py

```diff
@@ -0,0 +1,301 @@
+
+    @property
+    @_is_not_closed
+    def description(self):
+        """See also :py:meth: `Cursor.description`"""
+        if not self._cursor:
+            return None
+        columns_metadata = (
+            _metadata(name, metadata)
+            for name, metadata in self._cursor.columns().items()
+        )
+        columns_metadata = tuple(astuple(metadata) for metadata in columns_metadata)
+        columns_metadata = columns_metadata if columns_metadata != () else None
+        return columns_metadata
+
+    @property
+    @_is_not_closed
+    def rowcount(self):
+        """
+        See also :py:meth: `Cursor.rowcount`
+
+        Attention: This implementation of the rowcount attribute deviates slightly
+                   from what the dbapi2 requires.
+
+               Difference:
+
+                    If the rowcount of the last operation cannot be determined it will
+                    return 0.
+
+                    Expected by DBAPI2 = -1
+                    Actually returned  = 0
+        """
+        if not self._cursor:
+            return -1
+        return self._cursor.rowcount()
+
+    @_is_not_closed
+    def callproc(self, procname, parameters=None):
+        """See also :py:meth: `Cursor.callproc`"""
+        raise NotSupportedError("Optional and therefore not supported")
+
+    @_is_not_closed
+    def close(self):
+        """See also :py:meth: `Cursor.close`"""
+        self._is_closed = True
+        if not self._cursor:
+            return
+        self._cursor.close()
+
+    @_is_not_closed
+    def execute(self, operation, parameters=None):
+        """See also :py:meth: `Cursor.execute`"""
+        if parameters:
+            self.executemany(operation, [parameters])
+            return
+
+        connection = self._connection.connection
+        try:
+            self._cursor = connection.execute(operation)
+        except pyexasol.exceptions.ExaError as ex:
+            raise Error() from ex
+
+    @_is_not_closed
+    def executemany(self, operation, seq_of_parameters):
+        """See also :py:meth: `Cursor.executemany`"""
+        parameters = [
+            [_dbapi2pyexasol(p) for p in params] for params in seq_of_parameters
+        ]
+        connection = self._connection.connection
+        self._cursor = connection.cls_statement(connection, operation, prepare=True)
+        try:
+            self._cursor.execute_prepared(parameters)
+        except pyexasol.exceptions.ExaError as ex:
+            raise Error() from ex
+
+    def _convert(self, rows):
+        if rows is None:
+            return None
+
+        return tuple(self._convert_row(row) for row in rows)
+
+    def _convert_row(self, row):
+        if row is None:
+            return row
+
+        return tuple(
+            _pyexasol2dbapi(value, metadata)
+            for value, metadata in zip(row, self.description)
+        )
+
+    @_requires_result
+    @_is_not_closed
+    def fetchone(self):
```

Comment:
type hints, or would this break with dbapi

Reply:
I rather not, due to the fact that there 100% signature/typing in the pep.